### PR TITLE
SALTO-3786 Add change validator to check that guide is enabled when changing guide elements

### PR DESCRIPTION
**Pull request by idoamit8**
1. In this PR I added a change validator that validates that the user Cannot add instance when its associated brand has help center disabled.
2. I added test for this CV.
---

_Additional context for reviewer_
In Order to test the code manually - do as follows:
1. Create two environments in Zendesk.
2. Check\create two brands with the same name. 
3. Make sure that one brand has help center enabled and one brand has help center disabled. 
4. clone a sub guide element from the brand in the env with help center enabled to the env with help center disabled. 
5. try to deploy the new env. 
---
_Release Notes_: 
I added change validator to validate that it is not possible to add instance when its associated brand has help center disabled.

---
_User Notifications_: 
None

### CODE REVIEW

File: packages/zendesk-adapter/src/change_validators/guide_disabled.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    return []
+  }
+
+  const brandsWithGuideByInstanceId = await getBrandsWithGuideByInstanceId(relevantInstances)
+
+  return relevantInstances.filter(instance =>
+    instance.elemID.getFullName() in brandsWithGuideByInstanceId).map(instance => ({
+    elemID: instance.elemID,
+    severity: 'Error',
+    message: 'Cannot add instance because its associated brand has help center disabled.',
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/change_validators/guide_disabled.ts:56]: @tomermevorach can you review the `message` and `detailedMessage` ?

File: packages/zendesk-adapter/src/change_validators/guide_disabled.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  const relevantInstances = changes
+    .filter(isInstanceChange)
+    .filter(isAdditionChange)
+    .map(getChangeData)
+    .filter(instance => GUIDE_TYPES_TO_HANDLE_BY_BRAND.includes(instance.elemID.typeName))
+  if (_.isEmpty(relevantInstances)) {
+    return []
+  }
+
+  const brandsWithGuideByInstanceId = await getBrandsWithGuideByInstanceId(relevantInstances)
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/change_validators/guide_disabled.ts:50]:
```suggestion
  const brandsWithoutGuideByInstanceId = await getBrandsWithGuideByInstanceId(relevantInstances)
```

File: packages/zendesk-adapter/src/change_validators/guide_disabled.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+*/
+
+import { ChangeValidator, isInstanceChange, getChangeData, isAdditionChange, InstanceElement, isReferenceExpression } from '@salto-io/adapter-api'
+import _ from 'lodash'
+import { values as lowerDashValues } from '@salto-io/lowerdash'
+import { GUIDE_TYPES_TO_HANDLE_BY_BRAND } from '../config'
+
+const { isDefined } = lowerDashValues
+
+const getBrandsWithGuideByInstanceId = async (instances: InstanceElement[]):
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/change_validators/guide_disabled.ts:24]:
```suggestion
const getBrandsWithoutGuideByInstanceId = async (instances: InstanceElement[]):
```

File: packages/zendesk-adapter/src/change_validators/guide_disabled.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+Promise<Record<string, InstanceElement>> => {
+  const BrandsWithGuideByInstanceId = (await Promise.all(instances.map(async instance => {
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/change_validators/guide_disabled.ts:26]:
```suggestion
  const BrandsWithoutGuideByInstanceId = (await Promise.all(instances.map(async instance => {
```

File: packages/zendesk-adapter/src/change_validators/guide_disabled.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    const brandRef = instance.value.brand
+    if (!isReferenceExpression(brandRef)) {
+      return undefined
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/change_validators/guide_disabled.ts:29]: maybe add a log that it is not a reference expression?

File: packages/zendesk-adapter/src/change_validators/guide_disabled.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    .map(getChangeData)
+    .filter(instance => GUIDE_TYPES_TO_HANDLE_BY_BRAND.includes(instance.elemID.typeName))
+  if (_.isEmpty(relevantInstances)) {
+    return []
+  }
+
+  const brandsWithGuideByInstanceId = await getBrandsWithGuideByInstanceId(relevantInstances)
+
+  return relevantInstances.filter(instance =>
+    instance.elemID.getFullName() in brandsWithGuideByInstanceId).map(instance => ({
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/change_validators/guide_disabled.ts:53]:
```suggestion
return relevantInstances
.filter(instance => instance.elemID.getFullName() in brandsWithGuideByInstanceId)
.map(instance => ({
```

File: packages/zendesk-adapter/test/change_validators/guide_disabled.test.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    expect(changesErrors).toHaveLength(1)
+    expect(changesErrors).toEqual([{
+      elemID: categoryToBrandWithHelpCenterFalse.elemID,
+      severity: 'Error',
+      message: 'Cannot add instance because its associated brand has help center disabled.',
+      detailedMessage: `The brand "${brandWithHelpCenterFalse.elemID.name}" associated with this instance has help center disabled.`,
+    }])
+  })
+
+  it('should not return errors because the brand has help center', async () => {
```
Comment (edenhassid) [inline packages/zendesk-adapter/test/change_validators/guide_disabled.test.ts:70]: add tests for edge cases, for example: 
* no brand in the instance
* brand is not a reference expression
* no has_help_center

**Review by edenhassid — COMMENTED** [review]
good work! added a few comments :)  I think you can also use the `elementsSource` to get all the brands and have and elemID-brand Record, this way you won't have to use `getResolvedValue` for each change. notice that `getResolvedValue` can throw an error and in this case the program will crash.

File: packages/zendesk-adapter/src/change_validators/guide_disabled.ts [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+import { GUIDE_TYPES_TO_HANDLE_BY_BRAND } from '../config'
+
+const { isDefined } = lowerDashValues
+const log = logger(module)
+const { awu } = collections.asynciterable
+
+const getBrandsWithoutGuideByInstanceId = async (
+  instances: InstanceElement[], BrandsByBrandsId: Record<string, InstanceElement>):
+Promise<Record<string, InstanceElement>> => {
+  const BrandsWithoutGuideByInstanceId = (await Promise.all(instances.map(async instance => {
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/change_validators/guide_disabled.ts:31]: you don't need the `await Promise.all` and the `async` anymore cause you have the `BrandsByBrandsId`

File: packages/zendesk-adapter/src/change_validators/guide_disabled.ts [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    const brandRef = instance.value.brand
+    if (!isReferenceExpression(brandRef)) {
+      log.debug('brand is not a reference expression')
+      return undefined
+    }
+    const brand = BrandsByBrandsId[brandRef.elemID.getFullName()]
+    if (brand.value.has_help_center === false) {
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/change_validators/guide_disabled.ts:38]: brand can be undefined, I think you should check that before and if it is undefined add a log and return undefined

File: packages/zendesk-adapter/src/change_validators/guide_disabled.ts [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    .toArray()).map(instance => [instance.elemID.getFullName(), instance]))
+  if (_.isEmpty(BrandsByBrandsId)) {
+    return []
+  }
+
+  const brandsWithoutGuideByInstanceId = await getBrandsWithoutGuideByInstanceId(relevantInstances, BrandsByBrandsId)
+
+  return relevantInstances
+    .filter(instance =>
+      instance.elemID.getFullName() in brandsWithoutGuideByInstanceId).map(instance => ({
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/change_validators/guide_disabled.ts:73]:
```suggestion
.filter(instance =>instance.elemID.getFullName() in brandsWithoutGuideByInstanceId)
.map(instance => ({
```

**Review by edenhassid — COMMENTED** [review]
looks good! 🔥  added a few small comments

File: packages/zendesk-adapter/src/change_validators/guide_disabled.ts [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+      elemID: instance.elemID,
+      severity: 'Error',
+      message: 'Cannot add instance because its associated brand has help center disabled.',
+      detailedMessage: `The brand "${brandsWithoutGuideByInstanceId[instance.elemID.getFullName()].elemID.name}" associated with this instance has help center disabled.`,
+    }))
```
Comment (tomermevorach) [inline packages/zendesk-adapter/src/change_validators/guide_disabled.ts:78]: Can we change the message to:
"Cannot add this element because help center is not enabled for its associated brand"
And detailed message to:
"Please enable help center for brand {brand name} in order to add this element"

File: packages/zendesk-adapter/src/change_validators/guide_disabled.ts [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+import { collections, values as lowerDashValues } from '@salto-io/lowerdash'
+import { logger } from '@salto-io/logging'
+import { BRAND_TYPE_NAME } from '../constants'
+import { GUIDE_TYPES_TO_HANDLE_BY_BRAND } from '../config'
+
+const { isDefined } = lowerDashValues
+const log = logger(module)
+const { awu } = collections.asynciterable
+
+const getBrandsWithoutGuideByInstanceId = async (
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/change_validators/guide_disabled.ts:28]: you can remove the async :)

File: packages/zendesk-adapter/src/change_validators/guide_disabled.ts [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+    .filter(id => id.typeName === BRAND_TYPE_NAME)
+    .map(id => elementSource.get(id))
+    .filter(isInstanceElement)
+    .toArray())
+    .map(instance => [instance.elemID.getFullName(), instance]))
+  if (_.isEmpty(BrandsByBrandId)) {
+    return []
+  }
+
+  const brandsWithoutGuideByInstanceId = await getBrandsWithoutGuideByInstanceId(relevantInstances, BrandsByBrandId)
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/change_validators/guide_disabled.ts:74]: after you remove the async you can remove the await from here

**Review by edenhassid — APPROVED** [review]
looks good! 🚀  added two small comments, please fix that before you merge 😄
